lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,520 @@
+}) => string; // should return the domain of the origin
+
+export type TypeNormalisedInputRelyingPartyName = (input: {
+    request: BaseRequest | undefined;
+    userContext: UserContext;
+}) => string; // should return the app name
+
+export type TypeNormalisedInputGetOrigin = (input: { request: BaseRequest; userContext: UserContext }) => string; // should return the app name
+
+export type TypeInput = {
+    emailDelivery?: EmailDeliveryTypeInput<TypeWebauthnEmailDeliveryInput>;
+    relyingPartyId?: TypeInputRelyingPartyId;
+    relyingPartyName?: TypeInputRelyingPartyName;
+    getOrigin?: TypeInputGetOrigin;
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeInputRelyingPartyId =
+    | string
+    | ((input: { request: BaseRequest | undefined; userContext: UserContext }) => string);
+
+export type TypeInputRelyingPartyName =
+    | string
+    | ((input: { request: BaseRequest | undefined; userContext: UserContext }) => string);
+
+export type TypeInputGetOrigin = (input: { request: BaseRequest; userContext: UserContext }) => string;
+
+export type RecipeInterface = {
+    registerOptions(input: {
+        email: string;
+        relyingPartyId: string;
+        relyingPartyName: string;
+        origin: string;
+        timeout: number;
+        attestation: "none" | "indirect" | "direct" | "enterprise";
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<{
+        status: "OK";
+        webauthnGeneratedOptionsId: string;
+        rp: {
+            id: string;
+            name: string;
+        };
+        user: {
+            id: string;
+            name: string;
+            displayName: string;
+        };
+        challenge: string;
+        timeout: number;
+        excludeCredentials: {
+            id: string;
+            type: string;
+            transports: ("ble" | "hybrid" | "internal" | "nfc" | "usb")[];
+        }[];
+        attestation: "none" | "indirect" | "direct" | "enterprise";
+        pubKeyCredParams: {
+            alg: number;
+            type: string;
+        }[];
+        authenticatorSelection: {
+            requireResidentKey: boolean;
+            residentKey: "required" | "preferred" | "discouraged";
+            userVerification: "required" | "preferred" | "discouraged";
+        };
+    }>;
+
+    signInOptions(input: {
+        relyingPartyId: string;
+        origin: string;
+        timeout: number;
+        tenantId: string;
+        userContext: UserContext;
+    }): Promise<{
+        status: "OK";
+        webauthnGeneratedOptionsId: string;
+        challenge: string;
+        timeout: number;
+        userVerification: "required" | "preferred" | "discouraged";
```

Comment:
how is the value of userVerification defined?

Reply:
added them in the input and will add defaults for them. Done